src/wolff_producer.erl

```diff
@@ -446,35 +505,81 @@ handle_kafka_ack(#kpro_rsp{api = produce,
   [PartitionRsp] = kpro:find(partition_responses, TopicRsp),
   ErrorCode = kpro:find(error_code, PartitionRsp),
   BaseOffset = kpro:find(base_offset, PartitionRsp),
-  case ErrorCode =:= ?no_error of
-    true ->
-      do_handle_kafka_ack(Ref, BaseOffset, St);
-    false ->
+  case ErrorCode of
+    ?no_error ->
+      do_handle_kafka_ack(Ref, BaseOffset, St, normal);
+    ?message_too_large ->
+      do_handle_kafka_ack(Ref, BaseOffset, St, ?message_too_large);
+    _ ->
       #{topic := Topic, partition := Partition} = St,
       log_warn(Topic, Partition, "Produce response error-code = ~0p", [ErrorCode]),
       erlang:throw(ErrorCode)
   end.
 
-do_handle_kafka_ack(Ref, BaseOffset,
-                    #{sent_reqs := SentReqs,
-                      sent_reqs_count := SentReqsCount,
-                      pending_acks := PendingAcks,
-                      replayq := Q
-                     } = St) ->
+do_handle_kafka_ack(Ref, BaseOffset, #{sent_reqs := SentReqs } = St, Reason) ->
   case queue:peek(SentReqs) of
     {value, ?sent_req(#kpro_req{ref = Ref}, Q_AckRef, Calls)} ->
-      ok = replayq:ack(Q, Q_AckRef),
-      {{value, _}, NewSentReqs} = queue:out(SentReqs),
-      NewPendingAcks = evaluate_pending_ack_funs(PendingAcks, Calls, BaseOffset),
-      St#{sent_reqs := NewSentReqs,
-          sent_reqs_count := SentReqsCount - 1,
-          pending_acks := NewPendingAcks
-         };
+      %% this clause is kept only to be hot-upgrade safe
+      clear_sent_and_ack_callers(Q_AckRef, Calls, BaseOffset, St);
+    {value, ?sent_items(Ref, Items, Q_AckRef)} when Reason =:= ?message_too_large ->
+      Batch = get_batch_from_queue_items(Items),
+      case length(Batch) of
+        1 ->
+          %% This is a single message batch, but it's still too large
+          #{topic := Topic, partition := Partition} = St,
+          log_error(Topic, Partition, "One message is dropped due to single-message batch is too large!", []),
```

Comment:
Would it be worth throttling this kind of message, similar to what is done when logging discarded messages?

Reply:
the intention is to be very noisy